test/framework/src/main/java/org/elasticsearch/test/hamcrest/ElasticsearchAssertions.java

```diff
@@ -831,12 +832,12 @@ private static void assertListEquals(List<Object> expected, List<Object> actual)
      * for byte[] through Arrays.equals(byte[], byte[])
      */
     @SuppressWarnings("unchecked")
-    private static void assertObjectEquals(Object expected, Object actual) {
+    private static void assertObjectEquals(Object expected, Object actual, String path) {
         if (expected instanceof Map) {
             assertThat(actual, instanceOf(Map.class));
-            assertMapEquals((Map<String, Object>) expected, (Map<String, Object>) actual);
+            assertMapEquals((Map<String, Object>) expected, (Map<String, Object>) actual, path);
         } else if (expected instanceof List) {
-            assertListEquals((List<Object>) expected, (List<Object>) actual);
+            assertListEquals((List<Object>) expected, (List<Object>) actual, path);
```

Comment:
In assertToXContentEquivalent() we start with an empty path. Whenever assertMapEquals() finds an expected key, we append that to the path. Whenever an expected value or a map size differs, we additionally print the collected path information up until then. Its basically a recursion, "path" being the collector. the assertListEquals() method also needs the argument because we need to pass it on for other nested objects inside.